orttraining/orttraining/core/graph/graph_parser/pattern_graph.cc

```diff
@@ -0,0 +1,200 @@
+    auto* node = pattern_graph.GetNode(node_index);
+    if (strcmp(node->Name().c_str(), root_node.c_str()) == 0) {
+      pattern_root = node;
+      break;
+    }
+  }
+  if (!pattern_root) {
+    return Status(common::ONNXRUNTIME, common::FAIL, "Pattern root was not found.");
+  }
+
+  for (auto node_index : graph_topology_list) {
+    auto* node = target_graph.GetNode(node_index);
+    res.clear();
+    std::unordered_set<const Node*> graph_path, pattern_path;
+    std::unordered_map<const Node*, const Node*> path_map;
+    if (FindMatchRecursively(node, pattern_root, graph_path, pattern_path, path_map, res, target_graph)) {
+      res.push_back({node_index, pattern_root});
+      return Status::OK();
+    }
+  }
+
+  return Status(common::ONNXRUNTIME, common::FAIL, "No match for the target graph.");
+}
+
+bool DFSArgs(const Graph& graph, PatternGraph& pattern_graph, const std::unordered_map<std::string, const IArg*>& name_pargs_mapping,
+             ConstPointerContainer<std::vector<NodeArg*>>& p_args, ConstPointerContainer<std::vector<NodeArg*>>& t_args,
+             const std::unordered_map<std::string, std::unique_ptr<ArgCompareFunc>>& arg_constraints, const std::unique_ptr<ArgCompareFunc>& arg_default_constraint,
+             size_t p_arg_idx, std::unordered_set<const NodeArg*>& visited) {
+  if (p_arg_idx >= p_args.size()) return true;
+  auto p_arg = p_args[p_arg_idx];
+  PARSER_LOG << "Try to find match for arg " << p_arg->Name();
+  auto p_arg_name = p_arg->Name();
+  auto find_defined_arg = name_pargs_mapping.find(p_arg_name);
+  if (find_defined_arg == name_pargs_mapping.end()) return DFSArgs(graph, pattern_graph, name_pargs_mapping, p_args, t_args, arg_constraints, arg_default_constraint, p_arg_idx + 1, visited);
+  const IArg* p_arg_define = find_defined_arg->second;
+  auto func = arg_constraints.count(p_arg_name) > 0 ? arg_constraints.find(p_arg_name)->second.get() : arg_default_constraint.get();
+  for (auto t_arg : t_args) {
+    if (!visited.count(t_arg)) {
+      visited.insert(t_arg);
+      if ((*func)(t_arg, p_arg_define, graph, pattern_graph) &&
+          DFSArgs(graph, pattern_graph, name_pargs_mapping, p_args, t_args, arg_constraints, arg_default_constraint, p_arg_idx + 1, visited)) {
+        return true;
+      } else {
+        visited.erase(t_arg);
+        continue;
+      }
+    }
+  }
+  return false;
+};
+
+// let's assume that the graph and pattern do not have a circle.
+bool PatternGraph::FindMatchRecursively(const Node* g, const Node* p,
+                                        std::unordered_set<const Node*>& graph_path, std::unordered_set<const Node*>& pattern_path,
+                                        std::unordered_map<const Node*, const Node*>& path_map,
+                                        std::vector<PNN>& matched, const Graph& target) {
+  PARSER_LOG << "jump in. g: " << g->Name() << "  p: " << p->Name();
+
+  const std::string pnode_name = p->Name();
```

Comment:
BTW, during graph matching, will be the node output edge count be something we will check?